tokio/src/sync/batch_semaphore.rs

```diff
@@ -0,0 +1,672 @@
+//! atomic counter tracks the number of available permits. If the semaphore does
+//! not contain the required number of permits, the task attempting to acquire
+//! permits places its waker at the end of a queue. When new permits are made
+//! available (such as by releasing an initial acquisition), they are assigned
+//! to the task at the front of the queue, waking that task if its requested
+//! number of permits is met. 
+//!
+//! Because waiters are enqueued at the back of the linked list and dequeued
+//! from the front, the semaphore is fair. Tasks trying to acquire large numbers
+//! of permits at a time will always be woken eventually, even if many other
+//! tasks are acquiring smaller numbers of permits. This means that in a
+//! use-case like tokio's read-write lock, writers will not be starved by
+//! readers.
+use crate::loom::cell::CausalCell;
+use crate::loom::sync::{atomic::AtomicUsize, Mutex, MutexGuard};
+use crate::util::linked_list::{self, LinkedList};
+
+use std::{cmp, fmt};
+use std::future::Future;
+use std::marker::PhantomPinned;
+use std::pin::Pin;
+use std::ptr::NonNull;
+use std::sync::atomic::Ordering::*;
+use std::task::{Context, Poll, Waker};
+use std::task::Poll::*;
+
+/// An asynchronous counting semaphore which permits waiting on multiple permits at once.
+pub(crate) struct Semaphore {
+    waiters: Mutex<Waitlist>,
+    /// The current number of available permits in the semaphore.
+    permits: AtomicUsize,
+}
+
+struct Waitlist {
+    queue: LinkedList<Waiter>,
+    closed: bool,
+}
+
+/// Error returned by `Permit::try_acquire`.
+#[derive(Debug)]
+pub(crate) enum TryAcquireError {
+    Closed,
+    NoPermits,
+}
+/// Error returned by `Permit::poll_acquire`.
+#[derive(Debug)]
+pub(crate) struct AcquireError(());
+
+#[derive(Debug)]
+pub(crate) struct Permit {
+    state: PermitState,
+}
+
+pub(crate) struct Acquire<'a> {
+    node: Waiter,
+    semaphore: &'a Semaphore,
+    permit: &'a mut Permit,
+    num_permits: u16,
+}
+
+/// Permit state
+#[derive(Debug, Copy, Clone)]
+enum PermitState {
+    /// Currently waiting for permits to be made available and assigned to the
+    /// waiter.
+    Waiting(u16),
+
+    /// The number of acquired permits
+    Acquired(u16),
+}
+
+/// An entry in the wait queue.
+struct Waiter {
+    /// The current state of the waiter. 
+    /// 
+    /// This is either the number of remaining permits required by
+    /// the waiter, or a flag indicating that the waiter is not yet queued.
+    state: AtomicUsize,
+
+    /// The waker to notify the task awaiting permits.
+    /// 
+    /// # Safety
+    /// 
+    /// This may only be accessed while the wait queue is locked.
+    /// XXX: it would be nice if we could enforce this...
+    waker: CausalCell<Option<Waker>>,
+
+    /// Intrusive linked-list pointers.
+    /// 
+    /// # Safety
+    /// 
+    /// This may only be accessed while the wait queue is locked.
+    pointers: linked_list::Pointers<Waiter>,
```

Comment:
Yes, but it's hard to actually check this with `CausalCell`, since the `linked_list::Link` methods require _returning_ pointers to the `Pointers`.

 All of `CausalCell`'s APIs take closures, and use the closure to determine when the access ends. If we used `CausalCell` here, the `Link` implementation would have to return the pointers from inside that closure, and loom would consider the causal access to have "ended" while it was actually still ongoing. Given that, I wasn't sure it was worth using a `CausalCell`.

An alternative approach might be moving all the code that interacts with the pointers into `Waiter` methods that require a `MutexGuard<LinkedList<Waiter>>` reference, even when the head of the list isn't actually used. That would enforce that the pointers are only accessed while locked.

Reply:
The "fix" would be to work with `&self Pointers` and to use `CausalCell` inside `Pointers`?

Either way, it probably should be flagged for a follow up improvement since `Notify` has the same issue.

Reply:
I can add a comment?